src/v2/Artsy/Relay/middleware/cache/Cache.ts

```diff
@@ -0,0 +1,102 @@
+import { QueryResponseCache } from "relay-runtime"
+import createLogger from "v2/Utils/logger"
+import RelayQueryResponseCache from "relay-runtime/lib/network/RelayQueryResponseCache"
+import { isServer } from "lib/environment"
+
+const logger = createLogger("v2/Artsy/middleware/cache/Cache")
+
+export interface CacheConfig {
+  size: number
+  ttl: number
+}
+
+export class Cache {
+  cacheConfig: CacheConfig
+  relayCache: RelayQueryResponseCache
+
+  redisCache: {
+    del: (key: string) => Promise<void>
+    expire: (key: string, ttl: number) => Promise<void>
+    flushall: () => Promise<void>
+    get: (key: string) => Promise<string>
+    set: (key: string, value: string) => Promise<void>
+  }
+
+  constructor(cacheConfig: CacheConfig) {
+    this.cacheConfig = cacheConfig
+    this.initRelayCache()
+    this.initRedisCache()
+  }
+
+  initRelayCache() {
+    this.relayCache = new QueryResponseCache(this.cacheConfig)
+  }
+
+  initRedisCache() {
+    if (!isServer) {
+      return
+    }
+
+    const { promisify } = require("util")
+    const redis = require("redis")
+    const client = redis.createClient()
+
+    this.redisCache = {
+      del: promisify(client.del).bind(client),
+      expire: promisify(client.expire).bind(client),
+      flushall: promisify(client.flushall).bind(client),
+      get: promisify(client.get).bind(client),
+      set: promisify(client.set).bind(client),
+    }
+  }
+
+  getCacheKey(query, variables) {
+    const cacheKey = JSON.stringify({ query, variables })
+    return cacheKey
+  }
+
+  async get(queryId, variables) {
+    let cachedRes = this.relayCache.get(queryId, variables)
+
+    // No cache in relay store, check redis
+    if (isServer && !cachedRes) {
+      const cacheKey = this.getCacheKey(queryId, variables)
+
+      try {
```

Comment:
Similar to the approach in https://github.com/artsy/force/blob/f66806d89c01a536980d88e2693bedec302891b6/src/lib/middleware/pageCacheMiddleware.ts#L85 (which could be abstracted into a helper), I _think_ we should additionally guard against slow cache reads.

For whatever reason...we _occasionally_ see the cache get in a bad state where it can take hundreds of milliseconds to respond (quite literally the opposite of what you'd want). This has the effect of stalling and backing up Force, so there's that additional layer of `setTimeout` around the cache access to be able to reject it after the relatively low timeout threshold.

Reply:
+1

Reply:
This has been handled via Redis' retry options, which handles errors and so on